xds/internal/xdsclient/pool.go

```diff
@@ -0,0 +1,144 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsclient
+
+import (
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/internal/backoff"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/xds/bootstrap"
+)
+
+var (
+	// DefaultPool is the default pool for xds clients. It is created at the
+	// init time.
+	DefaultPool *Pool
+)
+
+// Pool represents pool of xds clients.
+type Pool struct {
+	mu      sync.Mutex
+	clients map[string]*clientRefCounted
+	config  *bootstrap.Config
+}
+
+func init() {
+	DefaultPool = &Pool{clients: make(map[string]*clientRefCounted)}
+}
+
+// NewPool creates a new xds client pool with the given bootstrap contents.
+func NewPool(bootstrapContents []byte) (*Pool, error) {
+	config, err := bootstrap.NewConfigForTesting(bootstrapContents)
+	if err != nil {
+		return nil, err
+	}
+	return &Pool{
+		clients: make(map[string]*clientRefCounted),
+		config:  config,
+	}, nil
+}
+
+// NewClient returns an xds client with the given name from the pool. If the
+// client doesn't already exist, it creates a new xds client and adds it to the
+// pool.
+//
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+func (p *Pool) NewClient(name string) (XDSClient, func(), error) {
+	config, err := bootstrap.GetConfiguration()
```

Comment:
What would be the reason to not use the `Pool.config` field xDS bootstrap configuration as that was the config passed in/parsed/created during the `NewPool` function above?

Reply:
NewClient is for the prod path which actually reads the bootstrap config from env vars and will only deal with default pool.

Reply:
This method *must* use the `config` field inside the `Pool`. There is no point in reading the bootstrap configuration from the env vars or the fallback bootstrap configuration. In fact, once we do that, we don't have a reason to have `Pool.NewClientForTesting`. Tests could as well use `pool.NewClient`.

What do you think?


> NewClient is for the prod path which actually reads the bootstrap config from env vars and will only deal with default pool

The bootstrap env vars should be read when the default pool is created at init time. And if bootstrap env vars are not set, the default pool should have a `nil` config and if `NewClient` is called without calling `SetFallbackBootstrapConfig`, i.e. `config` is `nil` when `NewClient` is called, then `NewClient` should fail with a useful error message.

Reply:
> The bootstrap env vars should be read when the default pool is created at init time. And if bootstrap env vars are not set, the default pool should have a nil config and if NewClient is called without calling SetFallbackBootstrapConfig, i.e. config is nil when NewClient is called, then NewClient should fail with a useful error message.

Done